### ISSUE TITLE
using predict() for regression using LSTM + time-distributed dense

### ISSUE THREAD

**Issue by alejandrok** [fchollet/keras#817]
Hi there!
We are working on performing audio source separation using a recurrent (LSTM) network architecture. We basically train the network using an audio magnitude spectrogram as an input (ie. vocals + instruments) and as an output, the corresponding vocals-only magnitude spectrogram. The spectrograms are normalized to [0, 1] so that we can use a sigmoid function in the output.

Based on the training and validation loss the network does a good job in learning this sort of transfer function we need and generalizing well (in theory, based on the validation loss) .We are using the MSE

SAMPLE OUTPUT FROM THE TRAINING STAGE:

Epoch 200/200
1059/1059 [==============================] - 6s - loss: 0.00044327 - val_loss: 0.00003266

but when it comes to testing it we are not getting a consistent output using predict. We suspect that we are doing something wrong in using predict(). We also don't know what the batch_size parameter in predict() does.

Any ideas about what could be wrong?

MODEL

freqBeans = 385 (number of freq coefficients in the spectrograms)
timestep = 32

model = Sequential()
model.add(LSTM(freqBeams, freqBeams, return_sequences=True))
model.add(LSTM(freqBeams, freqBeams, return_sequences=True))
model.add(LSTM(freqBeams, freqBeams, return_sequences=True))
model.add(TimeDistributedDense(freqBeams, freqBeams))
model.add(Activation('sigmoid'))
opt = SGD(lr=0.5, momentum=0.95, decay=1e-5, nesterov=False, clipnorm=.1)

AFTER TRAINING / AFTER SAVING THE WEIGHTS

model.load_weights("/storage/work/trainedModelKeras")
model.compile(loss='mean_squared_error', optimizer=opt)

TEST

prediction = model.predict(evalSample, batch_size=32)

![image](https://cloud.githubusercontent.com/assets/2191998/10414553/66f2148e-6f8f-11e5-888b-140f92e3c31d.png)

Based on the output prediction we suspect that there's something wrong not related to the network's performance. (Note that in the prediction spectrogram there is a significantly different vector every 32, which is exactly the timestep value. We suspect maybe predict() is throwing vectors instead of the full sequence or something like that. 

Any ideas about what could be wrong in the prediction phase?

**Comment by jfsantos** [comment]
This should work, I have code that does something similar (it even uses spectrograms as input and output). What are the dimensions of evalSample? It should have the same format as your dataset (i.e., `(num_samples, num_timesteps, num_features)`.

**Comment by alejandrok** [comment]
Thanks for your response. We have it working now! For some strange reason we were getting that output because of the activation function at the end of the network. We were using a relu so that should not have changed anything really; I wonder if there is any issue with Keras when an activation function is connected after a TimeDistributedDense layer. From the plot I posted it seems it is throwing one vector every 32 (the timestep value) instead of the whole sequence. We now removed it and we are getting the output we were expecting.
Thanks and good luck with your stuff. Are you working on source separation as well or what is that you're trying to solve?

**Comment by jfsantos** [comment]
I never used anything besides a sigmoid on the output layer for this, but I'll take a look into that. There should not be any issue since this kind of activation function is basically an elementwise operation. Thanks for the heads up! 

In my case, I'm working on speech enhancement, more specifically dereverberation.

**Comment by carlthome** [comment]
Glad to hear you got it working, @alejandrok. Would you mind sharing your fix as well? I'm interested in seeing how everything turned out.
